extensions/positron-run-app/src/api.ts

```diff
@@ -353,18 +353,27 @@ async function previewUrlInExecutionOutput(execution: vscode.TerminalShellExecut
 		return false;
 	}
 
-	// Convert the url to an external URI.
+	// Example: http://localhost:8500
 	const localBaseUri = vscode.Uri.parse(url.toString());
+
+	// Example: http://localhost:8500/url/path or http://localhost:8500
 	const localUri = urlPath ?
 		vscode.Uri.joinPath(localBaseUri, urlPath) : localBaseUri;
 
-	log.debug(`Viewing app at local uri: ${localUri} with external uri ${proxyInfo.externalUri.toString()}`);
+	// Determine the target origin based on whether the server is running in PWB.
+	const runningInPWB = process.env.RS_SERVER_URL ? true : false;
+	const targetOrigin = runningInPWB ? localBaseUri : localUri;
+	await proxyInfo.finishProxySetup(targetOrigin.toString(true));
+
+	// Example: http://localhost:8080/proxy/5678/url/path or http://localhost:8080/proxy/5678
+	const previewUri = runningInPWB && urlPath
+		? vscode.Uri.joinPath(proxyInfo.externalUri, urlPath)
+		: proxyInfo.externalUri;
 
-	// Finish the Positron proxy setup so that proxy middleware is hooked up.
-	await proxyInfo.finishProxySetup(localUri.toString());
+	log.debug(`Viewing app at local uri ${localUri} with external uri ${previewUri.toString(true)}`);
 
 	// Preview the external URI.
-	positron.window.previewUrl(proxyInfo.externalUri);
+	positron.window.previewUrl(previewUri);
```

Comment:
@seeM When running Fastapi in Positron on Workbench, sometimes the app url is previewed in the Viewer slightly too early, before the app server is ready. To get around this, a user can refresh the Viewer, but that is not ideal if it happens every time.

Do you have thoughts on how to get around this issue? Maybe we can ping the uri to check if it's responsive before displaying it in the Viewer?

Reply:
Maybe https://github.com/posit-dev/positron/issues/4276 is describing what I'm seeing?

Reply:
I assume that uvicorn would only print the URL to stdout once the server is ready to receive connections (I haven't confirmed that). Is the delay happening at the Positron Proxy layer? If so, maybe there's something we could expose from the Positron Proxy extension i.e. an `await`able or an event that we could listen to here?

Reply:
I looked into this a bit more and it does not seem to be in the Positron Proxy layer, as the issue still occurs even if we don't use the Positron Proxy.

For fastapi with uvicorn, we may want to wait for `Application startup complete` in the terminal output before we open the app at the url. We happen to be just a little quick to open the url/slow to startup the app on Workbench, such that the fastapi app is not quite ready yet.

Perhaps this would be a separate property in the terminal/debug config for some terminal text that indicates the app is ready, before we extract the app url and view it. I'll open a separate issue for this.

Reply:
Opened https://github.com/posit-dev/positron/issues/5187